src/liballoc/collections/btree/navigate.rs

```diff
@@ -120,56 +120,61 @@ impl<'a, K, V> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Ed
     }
 }
 
-impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
+impl<'a, K, V> Handle<NodeRef<marker::ValMut<'a>, K, V, marker::Leaf>, marker::Edge> {
     /// Moves the leaf edge handle to the next leaf edge and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the last one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
-    pub unsafe fn next_unchecked(&mut self) -> (&'a mut K, &'a mut V) {
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn next_unchecked(&mut self) -> (&'a K, &'a mut V) {
         unsafe {
             let kv = replace(self, |leaf_edge| {
                 let kv = leaf_edge.next_kv();
                 let kv = unwrap_unchecked(kv.ok());
                 (ptr::read(&kv).next_leaf_edge(), kv)
             });
-            // Doing the descend (and perhaps another move) invalidates the references
-            // returned by `into_kv_mut`, so we have to do this last.
-            kv.into_kv_mut()
+            // Doing this last is faster, according to benchmarks.
+            kv.into_kv_valmut()
         }
     }
 
     /// Moves the leaf edge handle to the previous leaf and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the first one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
-    pub unsafe fn next_back_unchecked(&mut self) -> (&'a mut K, &'a mut V) {
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn next_back_unchecked(&mut self) -> (&'a K, &'a mut V) {
         unsafe {
             let kv = replace(self, |leaf_edge| {
                 let kv = leaf_edge.next_back_kv();
                 let kv = unwrap_unchecked(kv.ok());
                 (ptr::read(&kv).next_back_leaf_edge(), kv)
             });
-            // Doing the descend (and perhaps another move) invalidates the references
-            // returned by `into_kv_mut`, so we have to do this last.
-            kv.into_kv_mut()
+            // Doing this last is faster, according to benchmarks.
+            kv.into_kv_valmut()
         }
     }
 }
 
+impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
+    /// Moves the leaf edge handle to the next leaf edge.
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn next_unchecked(&mut self) {
```

Comment:
Hm, seems odd to have a `next` function that does not return anything?

Reply:
Absolutely, that one is getting renamed. If I don't find anything similar in the libraries "advance_unchecked"? I hate that the other `next_unchecked` functions have the same name as in the map iterators, but that's what the reviewer at the time wanted. Elsewhere too Rusticians seem to love overloading names with subtly different meanings (like "node" in node.rs) and I shouldn't become complacent.